osidb/models.py

```diff
@@ -1518,6 +1536,71 @@ def __init__(self, *args, **kwargs):
     # a flaw can have many snippets, and a snippet can have many flaws
     flaws = models.ManyToManyField(Flaw, related_name="snippets", blank=True)
 
+    def convert_snippet_to_flaws(self) -> list:
+        """
+        Creates new flaws for all cve_ids in the snippet's content if a flaw with
+        the given cve_id does not exist, and links them together. If a flaw already
+        exists and does not contain the snippet of the current source, the snippet
+        will be linked to it.
+
+        Returns a list of newly created flaws.
+        """
+        created_flaws = []
+
+        for cve_id in self.content["cve_ids"]:
+            # cve_id cannot be an empty string when creating and saving a new flaw
+            cve_id = cve_id or None
+
+            if not cve_id or not Flaw.objects.filter(cve_id=cve_id):
+                flaw = self._create_flaw(cve_id=cve_id)
+                self.flaws.add(flaw)
+                created_flaws.append(flaw)
+            else:
+                flaw = Flaw.objects.filter(cve_id=cve_id).first()
+                if not flaw.snippets.filter(source=self.source):
+                    self.flaws.add(flaw)
+
+        return created_flaws
+
+    def _create_flaw(self, cve_id: Union[str, None]) -> Flaw:
+        """
+        Internal helper function to create a new flaw for the given `cve_id`. Any data
+        from the snippet's content that does not match the fields in Flaw will be ignored.
+
+        Returns a newly created flaw.
+        """
+        main_model = {}
+        related_models = {}
+
+        for key, value in self.content.items():
+            try:
+                field = Flaw._meta.get_field(key)
+
+                if field.is_relation:
+                    related_models[field.related_model] = value
+                elif field.model not in main_model:
+                    main_model[field.model] = {key: value}
+                else:
+                    main_model[field.model].update({key: value})
+            except FieldDoesNotExist:
+                # only "cve_ids" should end up here
+                pass
+
+        shared_acl = {"acl_read": self.acl_read, "acl_write": self.acl_write}
```

Comment:
Just wondering. How did we addressed the ACLs? Did we correctly set it to internal visibility when creating the Snippet? I cannot find any reference to it but maybe I am searching wrong.

Reply:
Yes, internal ACLs are set in `Snippet`'s `__init__`.

Reply:
I totally missed it :smile: It is there, all is good!